src/functional/ffront/foast_to_itir.py

```diff
@@ -31,61 +32,80 @@
 from functional.iterator import ir as itir
 
 
-class TypeKind(enum.Enum):
-    FIELD = 0
-    SCALAR = 1
-    UNKNOWN = 2
+def is_local_kind(symbol_type: ct.FieldType) -> bool:
+    assert isinstance(symbol_type, ct.FieldType)
+    if symbol_type.dims == ...:
+        return False
+    return any(dim.kind == DimensionKind.LOCAL for dim in symbol_type.dims)
 
 
-def resulting_type_kind(symbol_type: ct.SymbolType) -> TypeKind:
-    """
-    Determine whether the resulting type kind of ``symbol_type`` is scalar, field or unknown.
-
-    Useful for determining whether an expression of ``symbol_type`` should be treated as a
-    value or iterator expression during lowering.
-
-    Examples:
-    ---------
-    >>> resulting_type_kind(ct.DeferredSymbolType(constraint=None)).name
-    'UNKNOWN'
+class IteratorTypeKind(enum.Enum):
+    VALUE = 0
+    ITERATOR = 1
+    ENCAPSULATED_ITERATOR = 2
 
-    >>> resulting_type_kind(ct.ScalarType(kind=ct.ScalarKind.BOOL)).name
-    'SCALAR'
 
-    >>> resulting_type_kind(ct.TupleType(types=[ct.ScalarType(kind=ct.ScalarKind.FLOAT64)])).name
-    'SCALAR'
+def iterator_type_kind(
+    symbol_type: ct.ScalarType | ct.FieldType | ct.TupleType,
+) -> IteratorTypeKind:
     """
-    match symbol_type:
-        case ct.DeferredSymbolType(constraint=None):
-            return TypeKind.UNKNOWN
-        case ct.TupleType(types=subtypes):
-            if subtypes:
-                return resulting_type_kind(subtypes[0])
-
-    match type_info.type_class(symbol_type):
-        case ct.FieldType:
-            return TypeKind.FIELD
-        case ct.ScalarType:
-            return TypeKind.SCALAR
-
-    return TypeKind.UNKNOWN
-
-
-def can_be_value_or_iterator(symbol_type: ct.SymbolType):
-    return resulting_type_kind(symbol_type) is not TypeKind.UNKNOWN
-
-
-def is_field(expr: foast.Expr) -> bool:
-    return type_info.type_class(expr.type) is ct.FieldType
+    Return the kind of type on iterator level an foast expr of given type corresponds to.
```

Comment:
I stumbled over this sentence and had to read it twice, so I tried to rephrase it:
```suggestion
    Return the corresponding type kind (on iterator level) to a FOAST expression of the given SymbolType.
```